File: apps/web/src/lib/watchlist.ts
import type { Mode, Platform } from "./types";

export interface WatchFilters {
  market_hash_name: string;
  market: Platform | null;
  max_price_eur: string | null;
  max_float: string | null;
  paint_seeds: number[];
  doppler_phase: string | null;
}

export interface WatchRuleInput {
  name: string;
  enabled: boolean;
  filters: WatchFilters;
}

export interface WatchRule extends WatchRuleInput {
  id: string;
  mode: Mode;
  created_at: string;
  updated_at: string;
}

export interface WatchlistPage {
  mode: Mode;
  items: WatchRule[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

export function parsePaintSeeds(text: string): number[] {
  if (!text.trim()) return [];
  const parts = text.split(",").map((part) => part.trim());
  if (parts.length > 100 || parts.some((part) => !/^\d{1,4}$/.test(part) || Number(part) > 1000)) {
    throw new Error("Indiquez au plus 100 seeds de 0 à 1000, séparés par des virgules.");
  }
  return [...new Set(parts.map(Number))].sort((a, b) => a - b);
}

File: apps/api/app/schemas/watchlist.py
from datetime import datetime
from decimal import Decimal
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator

from app.schemas.api import Mode, Money, Platform

PaintSeed = Annotated[int, Field(strict=True, ge=0, le=1000)]
DopplerPhase = Literal[
    "Phase 1", "Phase 2", "Phase 3", "Phase 4", "Ruby", "Sapphire", "Black Pearl", "Emerald"
]


class WatchFilters(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    market_hash_name: str = Field(min_length=3, max_length=256, pattern=r"^[^\x00-\x1f\x7f]+$")
    market: Platform | None = None
    max_price_eur: Money | None = None
    max_float: Decimal | None = Field(default=None, ge=0, le=1, allow_inf_nan=False)
    paint_seeds: list[PaintSeed] = Field(default_factory=list, max_length=100)
    doppler_phase: DopplerPhase | None = None

    @field_validator("paint_seeds")
    @classmethod
    def distinct_seeds(cls, values: list[int]) -> list[int]:
        return sorted(set(values))


class WatchRuleInput(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    name: str = Field(min_length=1, max_length=128, pattern=r"^[^\x00-\x1f\x7f]+$")
    enabled: bool = Field(default=True, strict=True)
    filters: WatchFilters


class WatchRuleView(WatchRuleInput):
    model_config = ConfigDict(from_attributes=True, extra="forbid")

    id: str
    mode: Mode
    created_at: datetime
    updated_at: datetime


class WatchlistPage(BaseModel):
    mode: Mode
    items: list[WatchRuleView]
    total: int
    page: int
    page_size: int
    pages: int
